### PR TITLE
Anisotropic crack propagation along cleavage planes in crystalline materials

### LINKED ISSUE #14775: Anisotropic crack propagation along cleavage planes in crystalline materials
## Reason

Anisotropic crystalline materials may have several cleavage planes, some of them being weak. In this case, crack propagation is favored along the weak cleavage plane. The existing phase field damage model considers crack propagation in anisotropic materials only with respect to stiffness tensor. But the effect of cleavage plane needs to be considered as well. This material code will enhance the existing phase field model by including anisotropy due to presence of cleavage planes.

## Design

Input file will include a direction normal to the weak cleavage plane and a penalty parameter. The crack propagation along weak cleavage plane will be preferred as compared to other cleavage planes and is decided by the penalty parameter. According to the direction of weak cleavage plane (normal vector to cleavage plane), crack will propagate in a particular direction along the cleavage plane. The formulation is followed as given in attached paper Clayton & Knap 2015.
[Clayton & Knap 2015.pdf](https://github.com/idaholab/moose/files/4246862/Clayton.Knap.2015.pdf)

## Impact

This material code will enable current phase field model to be applicable to crystalline materials with several cleavage planes with varying weakness.

### PR DESCRIPTION
**Pull request by adandeka**
Enhancement - Considers crack propagation along weak cleavage planes in a material
Issue # 14775
closes #14775

### CODE REVIEW

**Comment by jiangwen84** [comment]
@adandeka You can visualize your documentation using the link https://mooseframework.inl.gov/docs/PRs/14824/site/source/kernels/ACInterfaceBetaPenalty.html

Looks like some of your equations are not correctly shown there. Please do not use your owner symbol shortcut. The reference should list in bib file. You can follow ComputeLinearElasticPFFractureStress.md

**Comment by jiangwen84** [comment]
@adandeka Could you derive your ACInterfaceBetaPenalty from ACInterface to avoid code duplication?

**Comment by adandeka** [comment]
@jiangwen84 
Yes, I think I can derive ACInterfaceBetaPenalty from ACInterface. Should I update the file and commit the change to this branch (feature_14775) ?
I will also look at the documentation file and update it.

**Comment by jiangwen84** [comment]
You can modify your existing PR and push to your origin, Github will automatically update your PR.

**Comment by adandeka** [comment]
@jiangwen84 I have updated all the files according to comments. Please let me know if I need to take any further action.

**Comment by jiangwen84** [comment]
@permcody Do you know what I can see this PR's documentation? 

The requested URL /docs/PRs/14824/site was not found on this server.

**Comment by permcody** [comment]
We've been exceedingly tight on space lately. Having 30+ PRs open is not sustainable on our current hardware. Jason has had no choice but to delete the docs for open PRs to make enough space. Hopefully we can migrate to another server soon.

File: modules/phase_field/src/kernels/ACInterfaceBetaPenalty.C [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+             _grad_test[_i][_qp](1) * _cleavage_plane_normal[1] +
+             _grad_test[_i][_qp](2) * _cleavage_plane_normal[2];
+
+  return _beta_penalty * _L[_qp] * _kappa[_qp] * ngradc * ngradpsi;
+}
+
+Real
+ACInterfaceBetaPenalty::computeQpResidual()
+{
+  return (1 + _beta_penalty) * _grad_u[_qp] * kappaNablaLPsi() - betaNablaPsi();
```
Comment (jiangwen84) [inline modules/phase_field/src/kernels/ACInterfaceBetaPenalty.C]: @adandeka Could you give the detailed derivation for the weak form of the second term? Add them to the documentation.

File: modules/phase_field/src/kernels/ACInterfaceBetaPenalty.C [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    _beta_penalty(getParam<Real>("beta_penalty")),
+    _cleavage_plane_normal(getParam<std::vector<Real>>("cleavage_plane_normal"))
+{
+}
+
+Real
+ACInterfaceBetaPenalty::betaNablaPsi()
+{
+  Real ngradc;
+  ngradc = _grad_u[_qp](0) * _cleavage_plane_normal[0] +
```
Comment (jiangwen84) [inline modules/phase_field/src/kernels/ACInterfaceBetaPenalty.C]: can be simplified as _grad_u[_qp] * _clevarge_plane_normal.

Same for the ngradpsi below.

File: modules/phase_field/include/kernels/ACInterfaceBetaPenalty.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+class ACInterfaceBetaPenalty;
+
+template <>
+InputParameters validParams<ACInterfaceBetaPenalty>();
+
+class ACInterfaceBetaPenalty : public ACInterface
+{
+public:
+  ACInterfaceBetaPenalty(const InputParameters & parameters);
```
Comment (jiangwen84) [inline modules/phase_field/include/kernels/ACInterfaceBetaPenalty.h]: Maybe use a different name for this class. ACInterfaceCleavageFracture.

File: modules/phase_field/include/kernels/ACInterfaceBetaPenalty.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+
+  /// term with beta penalty
+  Real betaNablaPsi();
+
+  // penalty for damage on planes not normal to the weak (favoured) cleavage
+  // plane (Clayton & Knap, 2015)
```
Comment (jiangwen84) [inline modules/phase_field/include/kernels/ACInterfaceBetaPenalty.h]: Use Doxygen comment format. See  http://www.doxygen.nl/manual/docblocks.html

Single line can use ///

File: modules/phase_field/doc/content/source/kernels/ACInterfaceBetaPenalty.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+added term is
+
+\begin{equation}
+\dfrac{\beta l_0}{2} \left( \boldsymbol{I} - \boldsymbol{n} \otimes \boldsymbol{n} \right) : \left( \nabla d \otimes \nabla d \right),
+\end{equation}
+
+Its weak form is
+
+\begin{equation}
+\left( - \beta l_0 \left( \nabla^2 d - \nabla d \otimes \nabla d : \nabla \nabla d \right) \right) \psi,
```
Comment (jiangwen84) [inline modules/phase_field/doc/content/source/kernels/ACInterfaceBetaPenalty.md]: Type here, \nabla d \otimes \nabla d  should be \nabla normal?

**Review by jiangwen84 — CHANGES_REQUESTED** [review]
Good job. Looks good. I have a few moments.

**Comment by jiangwen84** [comment]
@adandeka Would you have time to work on this? Thanks.

**Comment by adandeka** [comment]
@jiangwen84 
Thanks for checking. I am working on it. Will update the PR asap.

File: modules/phase_field/src/kernels/ACInterfaceBetaPenalty.C (hunk 2 above)
Comment (adandeka) [reply modules/phase_field/src/kernels/ACInterfaceBetaPenalty.C:35]: @jiangwen84 
_grad_u[_qp] * _clevarge_plane_normal  did not work.
It gives me error:
'const value_type {aka const libMesh::VectorValue<double>}' is not derived from 'const std::complex<_Tp>'

The declaration in the header file is   const std::vector<Real> _cleavage_plane_normal;
Does this multiplication needs any other type ?

File: modules/phase_field/doc/content/source/kernels/ACInterfaceCleavageFracture.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+\end{equation}
+
+The second term in above expression can be simplified as,
+
+\begin{equation}
+\begin{aligned}
+\int_{\Omega} \psi \left(\boldsymbol{n} \otimes \boldsymbol{n} : \nabla \nabla d \right) dV
+&= \int_{\Omega} \left( \nabla \cdot \left( \psi \boldsymbol{n} \left( \boldsymbol{n} \cdot \nabla d \right) \right) - \left( \boldsymbol{n} \cdot \nabla\psi \right) \left( \boldsymbol{n} \cdot \nabla d \right) - \left( \psi \nabla \cdot \boldsymbol{n} \right) \left( \boldsymbol{n} \cdot \nabla d \right) \right) dV
+\
+&= \int_{\Omega} \nabla \cdot \left( \psi \boldsymbol{n} \left( \boldsymbol{n} \cdot \nabla d \right) \right) dV - \int_{\Omega} \left( \boldsymbol{n} \cdot \nabla \psi \right) \left( \boldsymbol{n} \cdot \nabla d \right) dV - \int_{\Omega} \left( \psi \nabla \cdot \boldsymbol{n} \right) \left( \boldsymbol{n} \cdot \nabla d \right) dV
```
Comment (jiangwen84) [inline modules/phase_field/doc/content/source/kernels/ACInterfaceCleavageFracture.md:28]: The notation \boldsymbol{n} is a bit confusing. Can you use M for cleavage normal and n for the boundary normal?

And you might want to replace `\` with `\\` to put equations on multiple lines.

File: modules/phase_field/src/kernels/ACInterfaceCleavageFracture.C [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        _grad_phi[_j][_qp] * _dLdop[_qp] + _grad_u[_qp] * _phi[_j][_qp] * _d2Ldop2[_qp];
+
+    for (unsigned int i = 0; i < _n_args; ++i)
+      dgradL += (*_gradarg[i])[_qp] * _phi[_j][_qp] * (*_d2Ldargdop[i])[_qp];
+
+    dsum += (_kappa[_qp] * dgradL + _dkappadop[_qp] * _phi[_j][_qp] * gradL()) * _test[_i][_qp];
+  }
+
+  Real ngradc;
+  ngradc = _grad_u[_qp] * _cleavage_plane_normal;
```
Comment (jiangwen84) [inline modules/phase_field/src/kernels/ACInterfaceCleavageFracture.C:70]: One line is good. 

Real ngradc = _grad_u[_qp] * _cleavage_plane_normal;

same for ngradphi

File: modules/combined/test/tests/phase_field_fracture/crack2d_aniso_cleavage_plane.i [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+[Kernels]
+  [./ACbulk]
+    type = AllenCahn
+    variable = c
+    f_name = F
+  [../]
+  [./ACInterfaceCleavageFracture]
+    type = ACInterfaceCleavageFracture
+    variable = c
+    beta_penalty = 1
```
Comment (jiangwen84) [inline modules/combined/test/tests/phase_field_fracture/crack2d_aniso_cleavage_plane.i:51]: minor issue, is beta_penalty = 1 sufficient? I thought it has to be very large.

**Review by jiangwen84 — CHANGES_REQUESTED** [review]
Good job. Just minor changes.

File: modules/combined/test/tests/phase_field_fracture/crack2d_aniso_cleavage_plane.i (hunk 8 above)
Comment (adandeka) [reply modules/combined/test/tests/phase_field_fracture/crack2d_aniso_cleavage_plane.i:51]: beta=1 works well in cases where Gc ~ 2 J/m^2. If Gc is very high, then higher values of beta are required. The value in test input file is place holder.

File: modules/phase_field/src/kernels/ACInterfaceCleavageFracture.C (hunk 7 above)
Comment (adandeka) [reply modules/phase_field/src/kernels/ACInterfaceCleavageFracture.C:70]: ngradphi is already written in terms of _cleavage_plane_normal
    ngradphi = _grad_phi[_j][_qp] * _cleavage_plane_normal;

Should it be written differently ?

File: modules/phase_field/doc/content/source/kernels/ACInterfaceCleavageFracture.md (hunk 6 above)
Comment (adandeka) [reply modules/phase_field/doc/content/source/kernels/ACInterfaceCleavageFracture.md:28]: Thanks for the comment. I changed the notation. Now it is, M for cleavage plane normal and n for boundary normal. Also, the multiline equation now uses \\.

File: modules/phase_field/src/kernels/ACInterfaceCleavageFracture.C (hunk 7 above)
Comment (jiangwen84) [reply modules/phase_field/src/kernels/ACInterfaceCleavageFracture.C:70]: I mean you can combine two lines into one.

Comment (adandeka) [reply modules/phase_field/src/kernels/ACInterfaceCleavageFracture.C:70]: Thanks. I got it. Now, the two lines are combined.

**Comment by adandeka** [comment]
@jiangwen84 
Could you please help me understand this failed test (Mac Test) ? I did check the log but looked like some issue with libmesh.

**Comment by jiangwen84** [comment]
@adandeka can you first rebase your branch? git --rebase upstream devel?

**Comment by adandeka** [comment]
@jiangwen84 
Thanks for suggesting. I did rebase branch every time I updated files.
Now the Precheck fails in clang-format-check. I am applying clang-format package in atom but still it fails. Are the clang-format options for moose different than the standard package ?
Also, moosebuild used to generate a clang-format patch to fix any such issues but it does not generate it anymore.

**Comment by jiangwen84** [comment]
@permcody Do you know what causes precheck failure here? thanks.

**Comment by permcody** [comment]
Our bot's password expired. I just invalidated the target so it can run again.

**Comment by jiangwen84** [comment]
@permcody could you active the remaining tests? Thanks.

**Comment by jiangwen84** [comment]
@permcody I do not have any further comments. If those tests pass, could you please merge them. Thanks.

**Review by permcody — APPROVED** [review]

**Review by jiangwen84 — APPROVED** [review]

**Comment by adandeka** [comment]
@jiangwen84 
One quick question. Is there anything I need to do now ? Please let me know. Thanks.

**Comment by jiangwen84** [comment]
@adandeka No. Thanks for your great work. Sorry for taking so long. I look forward to your more contributions.

**Comment by jiangwen84** [comment]
@permcody could you merge it? Thanks.

**Pull request merged**